libyul/Utilities.cpp

```diff
@@ -86,56 +86,96 @@ std::string solidity::yul::reindent(std::string const& _code)
 	return out.str();
 }
 
-u256 solidity::yul::valueOfNumberLiteral(Literal const& _literal)
+LiteralValue solidity::yul::valueOfNumberLiteral(std::string_view const _literal)
 {
-	yulAssert(_literal.kind == LiteralKind::Number, "Expected number literal!");
-
-	static std::map<YulString, u256> numberCache;
-	static YulStringRepository::ResetCallback callback{[&] { numberCache.clear(); }};
+	return LiteralValue{LiteralValue::Data(_literal), std::string(_literal)};
+}
+LiteralValue solidity::yul::valueOfStringLiteral(std::string_view const _literal)
+{
+	std::string const s(_literal);
+	return LiteralValue{u256(h256(s, h256::FromBinary, h256::AlignLeft)), s};
+}
+LiteralValue solidity::yul::valueOfBuiltinStringLiteralArgument(std::string_view _literal)
+{
+	return LiteralValue{std::string(_literal)};
+}
+LiteralValue solidity::yul::valueOfBoolLiteral(std::string_view const _literal)
+{
+	if (_literal == "true")
+		return LiteralValue{1, "true"};
+	else if (_literal == "false")
+		return LiteralValue{0, "false"};
 
-	auto&& [it, isNew] = numberCache.try_emplace(_literal.value, 0);
-	if (isNew)
+	yulAssert(false, "Unexpected bool literal value!");
+}
+LiteralValue solidity::yul::valueOfLiteral(std::string_view const _literal, LiteralKind const& _kind, bool const _unlimitedLiteralArgument)
+{
+	switch (_kind)
 	{
-		std::string const& literalString = _literal.value.str();
-		yulAssert(isValidDecimal(literalString) || isValidHex(literalString), "Invalid number literal!");
-		it->second = u256(literalString);
+	case LiteralKind::Number:
+		return valueOfNumberLiteral(_literal);
+	case LiteralKind::Boolean:
+		return valueOfBoolLiteral(_literal);
+	case LiteralKind::String:
+		return _unlimitedLiteralArgument ? valueOfBuiltinStringLiteralArgument(_literal) : valueOfStringLiteral(_literal);
 	}
-	return it->second;
+	yulAssert(false);
```

Comment:
You can use this helper to mark the rest of a function as unreachable:
```suggestion
	util::unreachable();
```
We shouldn't even need to do that, but at some point GCC broke its unreachability heuristic and they haven't fixed that since (see #13102/#13087). Hopefully we'll be able to purge all those `unreachable()` calls some day.